### PR TITLE
core: add test to demonstrate canonicality confusion

### PR DESCRIPTION
**Pull request by holiman**
This PR adds a test where we

1. Have a chain [0 ... N .. X]
2. Reorg to shorter but heavier chain [0 ... N ... Y]
3. Then there should be no canon mapping for the block at height X

The headerchain insertion ensures that future 'stale' canon hashes are deleted, however, this is not done by blockchain when inserting blocks. This can lead to an inconsistent chain history, and would typically not appear on mainnet, but only in scenarios where 

- Forks/sidechains can be pretty long, 
- Difficulty varies wildly

The above two points especially concern Clique networks. 

Running this test yields
```
=== RUN   TestReorgToShorterRemovesCanonMapping
--- FAIL: TestReorgToShorterRemovesCanonMapping (0.04s)
    blockchain_test.go:1871: expected block to be gone: 80
    blockchain_test.go:1874: expected header to be gone: 80
FAIL
```

### CODE REVIEW

**Comment by holiman** [comment]
This PR now also contains

- A test which for header import (the fast-sync import mechanism). That one works correctly. 
- A fix for the flaw in blockchain.go. 

The underlying problem is most likely the cause behind the non-contiguous chain discussed in https://github.com/ethereum/go-ethereum/issues/19124#issuecomment-472690931 . The same thing has also been seen on our Rinkeby bootnode.

File: core/blockchain_test.go [hunk 1]
```diff
@@ -1810,3 +1811,127 @@ func TestPrunedImportSide(t *testing.T) {
...
+		t.Errorf("expected header to be gone: %v", headerByNum.Number.Uint64())
+	}
+
+}
+
+// TestReorgToShorterRemovesCanonMappingHeaderChain is the same scenario
+// as TestReorgToShorterRemovesCanonMapping, but applied on headerchain
+// imports -- that is, for fast sync
+func TestReorgToShorterRemovesCanonMappingHeaderChain(t *testing.T) {
+
```
Comment (karalabe) [inline core/blockchain_test.go]: Pls no empty line.

File: core/blockchain_test.go [hunk 2]
```diff
@@ -1810,3 +1811,127 @@ func TestPrunedImportSide(t *testing.T) {
...
+		t.Fatalf("head wrong, expected %x got %x", head.Hash(), got)
+	}
+	// We have now inserted a sidechain.
+	if blockByNum := chain.GetBlockByNumber(canonNum); blockByNum != nil {
+		t.Errorf("expected block to be gone: %v", blockByNum.NumberU64())
+	}
+	if headerByNum := chain.GetHeaderByNumber(canonNum); headerByNum != nil {
+		t.Errorf("expected header to be gone: %v", headerByNum.Number.Uint64())
+	}
+
```
Comment (karalabe) [inline core/blockchain_test.go]: Pls no empty line.

File: core/blockchain_test.go [hunk 3]
```diff
@@ -1810,3 +1811,127 @@ func TestPrunedImportSide(t *testing.T) {
...
```
Comment (karalabe) [inline core/blockchain_test.go]: Pls no empty line.

File: core/blockchain_test.go [hunk 4]
```diff
@@ -1810,3 +1811,127 @@ func TestPrunedImportSide(t *testing.T) {
...
+	}
+	return chain, longChain, heavyChain, nil
+}
+
+// TestReorgToShorterRemovesCanonMapping tests that if we
+// 1. Have a chain [0 ... N .. X]
+// 2. Reorg to shorter but heavier chain [0 ... N ... Y]
+// 3. Then there should be no canon mapping for the block at height X
+func TestReorgToShorterRemovesCanonMapping(t *testing.T) {
+
```
Comment (karalabe) [inline core/blockchain_test.go]: Pls no empty line.

**Pull request merged**